src/doorman/Fd.hpp

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include "Byte.hpp"
+
+#include <gsl/span>
+
+#include <utility>
+
+#include <sys/socket.h>
+#include <sys/uio.h>
+#include <unistd.h>
+
+class Fd {
+    int fd_;
+
+    void writev(gsl::span<const iovec> io_vecs) const;
+
+    static void *iovec_cast(const void *thing) { return const_cast<void *>(thing); }
+
+    template <typename T>
+    static iovec iovec_for(const T &some_object) {
+        static_assert(std::is_trivially_copyable_v<T>);
+        return {iovec_cast(&some_object), sizeof(some_object)};
+    }
+
+    static iovec iovec_for(gsl::span<const byte> bytes) { return {iovec_cast(bytes.data()), bytes.size_bytes()}; }
+    static iovec iovec_for(std::string_view string) { return {iovec_cast(string.data()), string.size()}; }
+
+    template <typename... Args>
+    static std::array<iovec, sizeof...(Args)> iovecs_for(Args &&... args) {
+        return {iovec_for(args)...};
+    }
+
+public:
+    Fd() : fd_(-1) {}
+    explicit Fd(int fd) noexcept : fd_(fd) {}
+    ~Fd() noexcept { close(); }
+    Fd(const Fd &) = delete;
+    Fd &operator=(const Fd &) = delete;
+    Fd(Fd &&other) noexcept : fd_(std::exchange(other.fd_, -1)) {}
+    Fd &operator=(Fd &&other) noexcept {
+        close();
+        fd_ = std::exchange(other.fd_, -1);
+        return *this;
+    }
+    [[nodiscard]] bool is_open() const noexcept { return fd_ >= 0; }
+    [[nodiscard]] int number() const {
+        if (!is_open())
+            throw std::runtime_error("Attempt to get handle for invalid file descriptor");
+        return fd_;
+    }
+    void close() {
+        if (is_open())
+            ::close(fd_);
+        fd_ = -1;
+    }
+    void write(gsl::span<const char> span) const { write(span.data(), span.size_bytes()); }
+    void write(gsl::span<const byte> span) const { write(span.data(), span.size_bytes()); }
+    void write(const void *data, size_t length) const;
+    template <typename T>
+    void write(const T &t) const {
+        write(&t, sizeof(t));
+    }
+    template <typename... Args>
+    void write_many(Args &&... args) const {
+        writev(iovecs_for(std::forward<Args>(args)...));
+    }
+
+    template <typename T>
+    T read_all() const {
+        T object;
+        read_all(&object, sizeof(object));
+        return object;
+    }
+    template <typename T>
+    void read_all(T &object) const {
+        read_all(&object, sizeof(object));
+    }
+    void read_all(void *data, size_t length) const;
```

Comment:
Do we need this function? It's very C and dangerous. What about passing in a std::vector<byte>& instead of void*?

Reply:
This function is used by the above function. We can't use a std::vector<> as not all data is being read into a vector: very commonly we're reading into a `Packet` object (using the above). I agree it's dangerous, and I'll see if I can make it at least `private` and only call with the `gsl::span<>` which covers essentially the same base as the `vector` you describe but works for any contiguous array of memory.

In fact, using `gsl::span` is a big improvement, so will take the time to use it more often! Thanks :)